torch/_jit_internal.py

```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
     You should not use this directly, it should only be used from the other
     createResolutionCallbackFrom* functions.
     """
+    def try_build_type(base_mod, subexp_mod):
+        try:
+            return base_mod[subexp_mod]
+        except TypeError:
+            return None
+
+    def subexpression(subexp, base_mod, module):
+        if ',' in subexp:
+            subexp = "".join(subexp.split())
+            splits = []
+            nest_level = 0
+            for i, c in enumerate(subexp):
+                if c == '[':
+                    nest_level += 1
+                elif c == ']':
+
+                    # fall back to c++ for unit test legacy behavior
+                    if nest_level == 0:
+                        return None
+                    nest_level -= 1
+                elif c == ',' and nest_level == 0:
+                    splits.append(i)
+
+            parts = []
+            if len(splits):
+                last_idx = 0
+                for idx in splits:
+                    parts.append(subexp[last_idx:idx])
+                    last_idx = idx + 1
+                parts.append(subexp[last_idx:])
+
+                types = [env(p, module) for p in parts]
+                if None in types:
+                    return None
+                return try_build_type(base_mod, tuple(types))
+
+            else:
+                subexp_mod = env(subexp, module)
+                if subexp_mod is None:
+                    return None
+                return try_build_type(base_mod, subexp_mod)
+        else:
+            # either base or subexp could include '.' or '['
+            if 'Tensor' == subexp:
+                subexp = 'torch.Tensor'
+            subexp_mod = env(subexp, module)
+            if subexp_mod is None:
+                return None
+            return try_build_type(base_mod, subexp_mod)
+
     def env(qualified_name, module):
         # We may need to resolve a qualified name, something like `torch.device`
         # or `a.b.c.d`. We first look up `torch` or `a` in the function's closed
         # over scope, then proceed to use the looked-up value to go down the
         # chain.
-        if '.' in qualified_name:
+        if '[' in qualified_name:
+            first, last = qualified_name.find('['), qualified_name.rfind(']')
+            base = qualified_name[:first]
+            base_mod = env(base, module)
+
+            # intentional bailout, e.g. base == "Tuple" but wasn't imported in Module
+            if not base_mod or isinstance(base_mod, torch.Tensor):
```

Comment:
oops, the part after 'or' was meant to be cleaned up. removing it after retesting..